hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/marker/DirectWriteMarkers.java

```diff
@@ -203,4 +238,49 @@ private Option<Path> create(Path markerPath, boolean checkIfExists) {
         + " in " + timer.endTimer() + " ms");
     return Option.of(markerPath);
   }
+
+  @Override
+  public Option<Path> createCompletionMarker(String partitionPath, String fileId, String instantTime, IOType type,
+                                                boolean ignoreExisting, Option<byte[]> serializedData)
+      throws HoodieException, IOException {
+    Path markerPath = getCompletionMarkerPath(partitionPath, fileId, instantTime, type);
+    if (!fs.exists(markerPath.getParent())) {
+      throw new HoodieException("Marker directory is absent." + markerPath.getParent());
+    }
+    try {
+      FSDataOutputStream outputStream = fs.create(markerPath, ignoreExisting);
+      if (serializedData.isPresent()) {
+        outputStream.writeLong(serializedData.get().length);
+        outputStream.write(serializedData.get());
+        outputStream.writeLong(generateChecksum(serializedData.get()));
+      }
+      outputStream.close();
+    } catch (IOException e) {
+      throw new HoodieException("Failed to create completed marker " + markerPath);
+    }
+    return Option.of(markerPath);
+  }
+
+  public static long generateChecksum(byte[] data) {
```

Comment:
nothing reuse from IOUtils? can we move this to some other utils class, more generic across code base